packages/editor-preview/src/browser/editor-preview-tree-decorator.ts

```diff
@@ -40,37 +40,48 @@ export class EditorPreviewTreeDecorator implements TreeDecorator, FrontendApplic
-
-    protected shell: ApplicationShell;
+    @inject(ApplicationShell) protected readonly shell: ApplicationShell;
     readonly id = 'theia-open-editors-file-decorator';
     protected decorationsMap = new Map<string, TreeDecoration.Data>();
 
     protected readonly decorationsChangedEmitter = new Emitter();
     readonly onDidChangeDecorations = this.decorationsChangedEmitter.event;
     protected readonly toDisposeOnDirtyChanged = new Map<string, Disposable>();
+    protected readonly toDisposeOnPreviewPinned = new Map<string, Disposable>();
 
     onDidInitializeLayout(app: FrontendApplication): void {
-        this.shell = app.shell;
         this.workspaceService.onWorkspaceChanged(() => {
             this.fireDidChangeDecorations((tree: Tree) => this.collectDecorators(tree));
         });
         this.workspaceService.onWorkspaceLocationChanged(() => {
             this.fireDidChangeDecorations((tree: Tree) => this.collectDecorators(tree));
         });
-        this.editorPreviewManager.onPreviewPinned(() => this.fireDidChangeDecorations((tree: Tree) => this.collectDecorators(tree)));
 
-        this.shell.onDidAddWidget(widget => this.registerSaveableListener(widget));
-        this.shell.onDidRemoveWidget(widget => this.toDisposeOnDirtyChanged.get(widget.id)?.dispose());
-        this.editorWidgets.forEach(widget => this.registerSaveableListener(widget));
+        this.shell.onDidAddWidget(widget => this.registerEditorListeners(widget));
+        this.shell.onDidRemoveWidget(widget => this.unregisterEditorListeners(widget));
+        this.editorWidgets.forEach(widget => this.registerEditorListeners(widget));
     }
 
-    protected registerSaveableListener(widget: Widget): void {
+    protected registerEditorListeners(widget: Widget): void {
         const saveable = Saveable.get(widget);
-        const isTrackableProvider = ApplicationShell.TrackableWidgetProvider.is(widget);
-        if (saveable && !isTrackableProvider) {
+        if (saveable && widget instanceof EditorPreviewWidget) {
             this.toDisposeOnDirtyChanged.set(widget.id, saveable.onDirtyChanged(() => {
                 this.fireDidChangeDecorations((tree: Tree) => this.collectDecorators(tree));
             }));
+            this.toDisposeOnPreviewPinned.set(widget.id, widget.onDidChangePreviewState(() => {
+                this.fireDidChangeDecorations((tree: Tree) => this.collectDecorators(tree));
+                this.toDisposeOnPreviewPinned.get(widget.id)?.dispose();
+                this.toDisposeOnDirtyChanged.delete(widget.id);
+            }));
+        }
+    }
+
+    protected unregisterEditorListeners(widget: Widget): void {
+        this.toDisposeOnDirtyChanged.get(widget.id)?.dispose();
+        this.toDisposeOnDirtyChanged.delete(widget.id);
+        if (widget instanceof EditorPreviewWidget && widget.isPreview) {
+            this.toDisposeOnPreviewPinned.get(widget.id)?.dispose();
+            this.toDisposeOnDirtyChanged.delete(widget.id);
```

Comment:
Mixed references here to `toDisposeOnPreviewPinned` and `toDisposeOnDirtyChanged`.

It looks like the logic above should ensure that you'll only have an entry in `toDisposeOnPreviewPinned` if the editor still `.isPreview`, but since there's no harm in trying to `.get` its `.id` or `.delete` it if it doesn't exist, I might remove the conditional anyway.

Reply:
Thanks for catching that, I've done both of these things